rtl/eclrtl/eclrtl.cpp

```diff
@@ -307,23 +307,48 @@ class RTLUnicodeConverter : public CInterface
 };
 
 typedef MapStringTo<RTLUnicodeConverter, char const *> MapStrToUnicodeConverter;
-MapStrToUnicodeConverter *unicodeConverterMap;
+static SafePointerArrayOf<MapStrToUnicodeConverter> *unicodeCoverterThreadMaps;
+static __thread MapStrToUnicodeConverter *unicodeConverterMap = NULL;
+static __thread ThreadTermFunc prevThreadTerminator = NULL;
 CriticalSection ucmCrit;
 
 MODULE_INIT(INIT_PRIORITY_STANDARD)
 {
-    unicodeConverterMap = new MapStrToUnicodeConverter;
+    unicodeCoverterThreadMaps = new SafePointerArrayOf<MapStrToUnicodeConverter>;
     return true;
 }
 MODULE_EXIT()
 {
-    delete unicodeConverterMap;
+    delete unicodeCoverterThreadMaps;
 }
 
+static void clearUnicodeConverterMap()
+{
+    {
+        CriticalBlock b(ucmCrit);
+        unicodeCoverterThreadMaps->zap(unicodeConverterMap);
+    }
+    if (prevThreadTerminator)
+        (*prevThreadTerminator)();
+}
 
 RTLUnicodeConverter * queryRTLUnicodeConverter(char const * codepage)
 {
-    CriticalBlock b(ucmCrit);
+    if (!unicodeConverterMap) // NB: one per thread, so no contention
+    {
+        unicodeConverterMap = new MapStrToUnicodeConverter(GetCurrentThreadId());
+
+        {
+            /* Store them in a global array for eventual disposal
+             * NB: in C++11, we would use TLS thread_local for the maps and would do away with the global map
+             */
+            CriticalBlock b(ucmCrit);
+            prevThreadTerminator = addThreadTermFunc(clearUnicodeConverterMap);
+        }
+        // Use thread terminator hook to clear them up on thread exit.
+        // NB: May need to revisit if not on a jlib Thread.
+        unicodeCoverterThreadMaps->append(unicodeConverterMap);
```

Comment:
You're right and that's the way it was before a final edit that added the block braces.. - whoops.